src/test/rgw/test_store_bench.cc

```diff
@@ -0,0 +1,281 @@
+
+	RGWNullFlusher flusher;
+	optional_yield y = null_yield;
+	RGWEnv env;
+	struct req_state s(cct, &env, 1);
+	int err = RGWUserAdminOp_User::create(&s, store, op_state, flusher, y);
+	if (err == -ERR_USER_EXIST) {
+		std::cout << "create user: " << Config::instance().uid << " already exists!"
+							<< std::endl;
+	}
+	std::cout << "create user: " << err << std::endl;
+}
+
+void list_users(rgw::sal::Store *store, common::CephContext *cct,
+								DoutPrefixProvider *dpp) {
+	RGWUserAdminOpState op_state(store);
+	JSONFormatter jf;
+	std::stringstream ss;
+	RGWStreamFlusher flusher(&jf, ss);
+	int err = RGWUserAdminOp_User::list(dpp, store, op_state, flusher);
+	std::cout << "list user: " << err << std::endl;
+	std::cout << "list " << ss.str() << std::endl;
+}
+
+void create_bucket() {
+	Config c = Config::instance();
+	RGWBucketInfo info;
+	RGWUserInfo owner;
+	obj_version objv;
+	rgw_placement_rule rule(
+			c.store->get_zone()->get_zonegroup().get_default_placement_name(),
+			RGW_STORAGE_CLASS_STANDARD);
+	std::map<std::string, bufferlist> attrs;
+	owner.user_id.id = Config::instance().uid.id;
+
+	// idk what this is
+	objv.ver = 2;
+	objv.tag = "write_tag";
+	rule.name = "rule1";
+	rule.storage_class = "sc1";
+	auto user = c.store->get_user(Config::instance().uid);
+	rgw_placement_rule r(
+			c.store->get_zone()->get_zonegroup().get_default_placement_name(),
+			RGW_STORAGE_CLASS_STANDARD);
+	owner.default_placement = r;
+	RGWQuotaInfo qi;
+	auto zone = c.store->get_zone();
+	RGWAccessControlPolicy cp;
+	RGWEnv env;
+	req_info req_i(c.cct, &env);
+	std::unique_ptr<rgw::sal::Bucket> bucket;
+	std::string swift_var_location;
+	bool existed = false;
+	int ret =
+			user->create_bucket(c.dpp, c.bucket, zone->get_zonegroup().get_id(), r,
+													swift_var_location, &qi, cp, attrs, info, objv, false,
+													false, &existed, req_i, &bucket, null_yield);
+	std::cout << "create bucket: " << ret << std::endl;
+	std::cout << "existed: " << existed << std::endl;
+}
+
+void list_buckets() {
+	auto c = Config::instance();
+	auto user = c.store->get_user(c.uid);
+	rgw::sal::BucketList buckets;
+	std::string marker, end_marker;
+	user->list_buckets(c.dpp, marker, end_marker, 10, 0, buckets, null_yield);
+	std::cout << "Buckets: ";
+	for (auto &[name, bucket] : buckets.get_buckets()) {
+		std::cout << name << " ";
+	}
+	std::cout << std::endl;
+}
+
+static bufferlist *bl = nullptr;
+// bl->length() doesn't work?????
+static uint32_t bufferlist_length = 0;
+
+uint64_t create_object(rgw_obj_key key, uint32_t obj_size) {
+	// For now let's have one bufferlist as append might be too expensive
+	if (bl == nullptr) {
+		bl = new bufferlist();
+		bl->append_zero(obj_size);
+		bufferlist_length = obj_size;
+	}
+	if (bufferlist_length != obj_size) {
+		bl->clear();
+		bl->append_zero(obj_size);
+		bufferlist_length = obj_size;
+	}
+
+	std::unique_ptr<rgw::sal::Object> obj(
+			new rgw::sal::RadosObject(((rgw::sal::RadosStore *)CONF.store), key));
```

Comment:
Please don't create a RadosObject, it's a layer violation.  Instead, use store->get_object(key), which creates the correct type of Object without doing a lookup.

Reply:
Ahh I get it now.

Reply:
Can I ask why is it that way? Why not have `get_object` and `lookup_object` instead? Anyways, should have read that before.

Reply:
get_object() just creates the object instance in memory.  Object::get_obj_state() and Object::get_obj_attrs() load the state and attributes (respectively) from the backing store.  The reason for this is that Object frequently refers to an object that has not yet been created (or has already been deleted) in the backing store.